src/parser/command.rs

```diff
@@ -213,6 +220,42 @@ pub fn parse_defframe<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instructio
     ))
 }
 
+/// Parse the contents of a `DEFGATE` instruction.
+pub fn parse_defgate<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruction> {
+    let (input, name) = token!(Identifier(v))(input)?;
+    let (input, parameters) = opt(delimited(
+        token!(LParenthesis),
+        separated_list1(token!(Comma), token!(Variable(v))),
+        token!(RParenthesis),
+    ))(input)?;
+    let (input, gate_type) = opt(preceded(
+        token!(As),
+        alt((
+            map(token!(Matrix), |()| GateType::Matrix),
+            map(token!(Permutation), |()| GateType::Permutation),
+        )),
+    ))(input)?;
+    let (input, _) = token!(Colon)(input)?;
+
+    let gate_type = gate_type.unwrap_or(GateType::Matrix);
+    let (input, specification) = match gate_type {
+        GateType::Matrix => {
+            parse_matrix(input).map(|(input, matrix)| (input, GateSpecification::Matrix(matrix)))?
```

Comment:
The `map` combinator from `nom` allows you to map just the output without worrying about the leftover input. I think you could rewrite this line as:

```
map(parse_matrix, GateSpecification::Matrix)(input)
```